pkg/kv/kvserver/multi_sst_writer.go

```diff
@@ -382,17 +382,43 @@ func (msstw *multiSSTWriter) PutInternalRangeDelete(ctx context.Context, start,
 func (msstw *multiSSTWriter) PutInternalRangeKey(
 	ctx context.Context, start, end []byte, key rangekey.Key,
 ) error {
-	decodedStart, decodedEnd, err := decodeRangeStartEnd(start, end)
-	if err != nil {
-		return err
+	return msstw.putRangeKey(ctx, storage.EngineKeyRange{}, [2][]byte{start, end}, key)
+}
+
+func (msstw *multiSSTWriter) putRangeKey(
```

Comment:
I have mixed feelings about this one. The amount of duplication in the old code wasn't bad, only needed to call `rolloverSST` and `Frag.Add`. The `rolloverSST` is done by all other similar funcs, so it's not a bad duplication, just following the pattern. So only `Frag.Add()` would be "badly" duplicated. Seems ok, if compared to extra complexity in this normalization commit.

Alternatively, maybe this is fixable further down the stack, so that e.g. `rolloverSST` takes a different type. Haven't thought about this much.

Reply:
Yeah, you're right, the reason for deduplicating originally was that a bunch of code was duplicated (and the important code was missing in the copy that mattered), but with the movement of stats into the fragmenters it's less the case now. I'll stop playing code golf here for now, though, it's clear what the code does and it's as obviously correct as maintaining two versions of the code. Could've ended up the other way in a parallel universe.